File: course-schedule.py
# TC: O(V+E)
# SC: O(V+E)


# intuition: BFS
# create an adjacency map i.e. who are the neighbors of a given node (outgoing) so that you can tarverse through these nodes. (Independent nodes as Keys & dependent nodes are their values)
# create indegrees array. Note count of incoming arrows for a given node with idx of the array = node value
# maintain a queue of independent nodes for BFS
# Put ALL independent nodes in the queue.
# Iterate through the queue, for each independent node, find its neighbors. Decrement their counts in inDegrees array that means we have traverse through those node. If any value becomes 0 i.e. node becomes independent, add that node to queue. Keep doing this until queue is not empty.

# Once q is empty, loop through indegrees array. If all values are not 0, that means some course was not taken return false. Else return True
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """

        # size = len(prerequisites) + 1
        # print(size)
        inDeg = [0 for i in range(0, numCourses)]
        q = []
        adjMap = {}

        for i in prerequisites:
            # i is a vector or the edge in this case

            inA = i[0]
            outA = i[1]
            inDeg[inA] += 1
            if outA not in adjMap:
                adjMap[outA] = []
            adjMap[outA].append(inA)

        print(adjMap)

        for i in range(numCourses):
            if inDeg[i] == 0:
                q.append(i)

        while len(q) > 0:
            independentCourse = q.pop(0)
            print("pop", independentCourse)
            print("queue", q)
            # print(adjMap[independentCourse])
            if independentCourse in adjMap:
                dependentCourses = adjMap[independentCourse]
                for course in dependentCourses:
                    inDeg[course] -= 1
                    if inDeg[course] == 0:
                        q.append(course)

        for i in range(len(inDeg)):
            if inDeg[i] != 0:
                return False

        return True

File: Course_Schedule.java
//Time Complexity: O(V+E)
//TimeComplexity : O(n)

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        HashMap<Integer, List<Integer>> map = new HashMap<>();
        Queue<Integer> queue = new LinkedList<>();
        int []indegrees = new int[numCourses];
        for(int []subArray : prerequisites){
            int dependent = subArray[0];
            int independent = subArray[1];
            indegrees[dependent]++;
            if(!map.containsKey(independent)) {
                map.put(independent, new ArrayList<>());
            }
            map.get(independent).add(dependent);
        }
       int countfinish=0;
        for(int i =0; i<numCourses; i++) {
            if(indegrees[i] == 0) {
                queue.add(i);
                countfinish++;
            }
        }
        if(queue.isEmpty()) {
            return false;
        }
        while(!queue.isEmpty()) {
            int current = queue.poll();
            List<Integer> res = map.get(current);
            if(res == null) {
                continue;
            }
            for(int r : res) {
                indegrees[r]--;
                if(indegrees[r] == 0) {
                    queue.add(r);
                    countfinish++;
                    if(countfinish == numCourses) {
                        return true;
                    }
                }
                
            }
        }
      if(countfinish == numCourses) {
                    return true;
                }
                return false;
    }
}